NitroxServer/GameLogic/Bases/BaseManager.cs

```diff
@@ -9,153 +9,169 @@ namespace NitroxServer.GameLogic.Bases
+        // the building layout is handled by the client on initial loading.  
+        private List<BasePiece> allBasePieces;
 
-        public BaseManager(List<BasePiece> partiallyConstructedPieces, List<BasePiece> completedBasePieceHistory)
+        public BaseManager(List<BasePiece> allBasePieces)
         {
-            this.completedBasePieceHistory = completedBasePieceHistory;
-            partiallyConstructedPiecesById = partiallyConstructedPieces.ToDictionary(piece => piece.Id);
+            this.allBasePieces = allBasePieces;
         }
 
-        public List<BasePiece> GetCompletedBasePieceHistory()
+        public List<BasePiece> GetAllBasePieces()
         {
-            lock(completedBasePieceHistory)
+            lock (allBasePieces)
             {
-                return new List<BasePiece>(completedBasePieceHistory);
+                return new List<BasePiece>(allBasePieces);
             }
         }
 
-        public List<BasePiece> GetPartiallyConstructedPieces()
+        public void BasePieceConstructionBegin(BasePiece basePiece)
         {
-            lock (partiallyConstructedPiecesById)
-            {
-                return new List<BasePiece>(partiallyConstructedPiecesById.Values);
-            }
-        }
 
-        public void AddBasePiece(BasePiece basePiece)
-        {
-            lock (partiallyConstructedPiecesById)
+#if TRACE && BUILDING
+            NitroxModel.Logger.Log.Debug("BasePieceConstructionBegin - id: " + basePiece.Id + " - basePiece: " + basePiece);
+#endif
+
+            lock (allBasePieces)
             {
-                partiallyConstructedPiecesById.Add(basePiece.Id, basePiece);
+                // Subnautica changes the order of basepieces if there are Layout-Changes on an existing base, e.g. by adding pieces or splitting a base apart
+                // in two bases by removing connector-tubes. This changes the order in which BasePieces are need to be generated by Nitrox on client side loading.
+                if (allBasePieces.Count > 0)
+                {
+                    int nextBuildIndex = allBasePieces.Max(piece => piece.BuildIndex) + 1;
+                    basePiece.BuildIndex = nextBuildIndex;
+                }
+
+                allBasePieces.Add(basePiece);
             }
         }
 
         public void BasePieceConstructionAmountChanged(NitroxId id, float constructionAmount)
         {
+
+#if TRACE && BUILDING
+            NitroxModel.Logger.Log.Debug("BasePieceConstructionAmount - id: " + id + " - constructionAmount: " + constructionAmount);
+#endif
+
             BasePiece basePiece;
 
-            lock (partiallyConstructedPiecesById)
+            lock (allBasePieces)
             {
-                if (partiallyConstructedPiecesById.TryGetValue(id, out basePiece))
+                basePiece = allBasePieces.Find(piece => piece.Id == id);
```

Comment:
It looks like almost all of the access patterns are looking up by id.  Should we index to a dictionary similar to how partiallyConstructedPiecesById used to be done in the constructor?

Reply:
Using a dictionary might also be possible. But this would lead to more code in handling the buildindex or other constellations where we may need to find a group of items based on a parameter (e.g. base splitting could cause a group of items to update their parents). List gives here more flexibility to not explicitly make use of for each loops.
In concerns of performance i'm with you, that the list operations are a little bit slower than the dictionarys, when the only use case is to get the value to a specific key. When you need to iterate through all key/value pairs of a dictionary the performance difference is not so big anymore. Also the difference is in magnitue of microseconds for the number of entrys that are handled here (maybe about 500 to 1000 items for multiple bases in full playthrough?)  and the methods are called only at specific situations in the game (maybe about 500 calls in a game session?). 
For updating movement infos that are need to be handled permanently i would switch, but here i would prefer to stay with the easier coding way and the more flexibility.